### PR TITLE
Add e2e tests for Drag-and-Drop in the inserter

### LINKED ISSUE #31356: e2e Tests: Add test for inserter drag-and-drop
GB 10.5.0 had a small regression related to the inserter's drag-and-drop feature (see #31319). Let's try adding some e2e test coverage for that feature (maybe to https://github.com/WordPress/gutenberg/blob/cec7a6817ac2ec37e1e7a3b314fee0b909e7bd7d/packages/e2e-tests/specs/editor/various/inserting-blocks.test.js?).

Note that it has been previously [thought to be impossible](https://github.com/WordPress/gutenberg/pull/31319#issuecomment-829042542) to cover drag-and-drop behavior with an e2e test. However, there is some prior art for blocks' drag-and-drop feature, see #26869.

### PR DESCRIPTION
**Pull request by kevin940726**
## What?

Close https://github.com/WordPress/gutenberg/issues/31356. Add drag-and-drop e2e tests for the global inserter.

This PR also tries to add cross-browser testing with firefox and webkit, powered by Playwright.

## Why?

Because more tests = better! It will also make a potential refactor of the drag-and-drop feature easier to implement.

## How?

By default, e2e tests are only run in chromium. We can enable testing in firefox and webkit by appending the `@firefox` or `@webkit` tag to the test title. The tag also works in `test.describe` and will apply to all the nested tests. We can append `-firefox`, `-webkit`, or `-chromium` to individually disable testing in some browsers.

## Testing Instructions

```sh
npm run test:e2e:playwright -- inserting-blocks
```

CI should pass.

### CODE REVIEW

File: packages/block-editor/src/components/inserter-list-item/style.scss [hunk 1]
```diff
@@ -49,6 +49,8 @@
 				border-radius: $radius-block-ui;
 				opacity: 0.04;
 				background: var(--wp-admin-theme-color);
+				// This fixes drag-and-drop in Firefox.
+				pointer-events: none;
```
Comment (kevin940726) [inline packages/block-editor/src/components/inserter-list-item/style.scss:53]: Reference: https://github.com/WordPress/gutenberg/pull/44711#discussion_r992510624.

File: .github/workflows/end2end-test-playwright.yml [hunk 2]
```diff
@@ -39,22 +39,22 @@ jobs:
...
-                  npm run test:e2e:playwright
+                  xvfb-run --auto-servernum --server-args="-screen 0 1280x960x24" -- npm run test:e2e:playwright
 
             - name: Archive debug artifacts (screenshots, traces)
               uses: actions/upload-artifact@3cea5372237819ed00197afe530f5a7ea3e805c8 # v3.1.0
               if: always()
               with:
                   name: failures-artifacts
-                  path: artifacts
+                  path: artifacts/test-results
```
Comment (kevin940726) [inline .github/workflows/end2end-test-playwright.yml]: This is a long-standing bug that kept uploading storage-state to artifacts when there were no failures 😓.

File: test/e2e/playwright.config.ts [hunk 3]
```diff
@@ -54,6 +55,30 @@ const config: PlaywrightTestConfig = {
...
+				...devices[ 'Desktop Safari' ],
+				/**
+				 * Headless webkit won't receive dataTransfer with custom types in the
+				 * drop event on Linux. The solution is to use `xvfb-run` to run the tests.
+				 * ```sh
+				 * xvfb-run npm run test:e2e:playwright
+				 * ```
+				 * See `.github/workflows/end2end-test-playwright.yml` for advanced usages.
+				 */
+				headless: os.type() !== 'Linux',
```
Comment (kevin940726) [inline test/e2e/playwright.config.ts:72]: This is kinda unfortunate for linux developers. I'll try to raise an issue upstream. Maybe we should consider only running chromium tests when developing locally?

Comment (kevin940726) [reply test/e2e/playwright.config.ts:72]: Opened an issue here: https://github.com/microsoft/playwright/issues/18013.

**Comment by talldan** [comment]
Would be good to get CI passing in this PR. I think the problem causing them to fail has been solved in `trunk`, so a rebase should fix things.

**Comment by aaronrobertshaw** [comment]
Nice job working through the kinks in getting the cross-browser testing in place @kevin940726 👌 

Everything tests well for me, changes make sense, and CI is passing. I'm inclined to approve this PR but think it might be better to get the official green light from someone with more expertise in this area.

File: .github/workflows/end2end-test-playwright.yml (hunk 2 above)
Comment (talldan) [reply .github/workflows/end2end-test-playwright.yml]: Might be good to move this into a separate PR to keep the changes atomic.

File: packages/block-editor/src/components/inserter-list-item/style.scss (hunk 1 above)
Comment (talldan) [inline packages/block-editor/src/components/inserter-list-item/style.scss:53]: Here as well, this could be a separate PR.

File: test/e2e/playwright.config.ts [hunk 4]
```diff
@@ -54,6 +55,30 @@ const config: PlaywrightTestConfig = {
 		{
 			name: 'chromium',
 			use: { ...devices[ 'Desktop Chrome' ] },
+			grepInvert: /-chromium/,
+		},
+		{
+			name: 'webkit',
+			use: {
+				...devices[ 'Desktop Safari' ],
```
Comment (talldan) [inline test/e2e/playwright.config.ts:63]: So this isn't exactly Safari, but a Webkit based browser? Interesting.

File: .github/workflows/end2end-test-playwright.yml [hunk 5]
```diff
@@ -39,22 +39,22 @@ jobs:
 
             - name: Install Playwright dependencies
               run: |
-                  npx playwright install chromium --with-deps
+                  npx playwright install chromium firefox webkit --with-deps
```
Comment (talldan) [inline .github/workflows/end2end-test-playwright.yml:42]: It's a shame there isn't a better summary of the browsers in the test output. Dividing the test results by browser would be nice. In CI we can at least see that two tests were skipped (the firefox ones), so that validates that the tests are running in FF.

I suppose it can't be helped if that's how playwright does it. 😄 

Surprisingly it's also not clear in the more verbose mode when run locally:
```
  ✓  1 …Inserting blocks (@firefox, @webkit) › inserts blocks by dragging and dropping from the global inserter (5s)
  ✓  2 …erting blocks (@firefox, @webkit) › cancels dragging blocks from the global inserter by pressing Escape (5s)
  ✓  3 …serting blocks (@firefox, @webkit) › inserts patterns by dragging and dropping from the global inserter (3s)
  ✓  4 …ting blocks (@firefox, @webkit) › cancels dragging patterns from the global inserter by pressing Escape (3s)
  ✓  5 …Inserting blocks (@firefox, @webkit) › inserts blocks by dragging and dropping from the global inserter (7s)
  ✓  6 …erting blocks (@firefox, @webkit) › cancels dragging blocks from the global inserter by pressing Escape (5s)
  ✓  7 …serting blocks (@firefox, @webkit) › inserts patterns by dragging and dropping from the global inserter (4s)
  ```

File: test/e2e/specs/editor/various/inserting-blocks.spec.js [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+			'role=region[name="Block Library"i] >> role=searchbox[name="Search for blocks and patterns"i]',
+			PATTERN_NAME
+		);
+
+		await page.hover(
+			`role=listbox[name="Block Patterns"i] >> role=option[name="${ PATTERN_NAME }"i]`
+		);
+
+		// FIXME: I think we should show the indicator when hovering on patterns as well?
+		// await expect( insertingBlocksUtils.indicator ).toBeVisible();
```
Comment (talldan) [inline test/e2e/specs/editor/various/inserting-blocks.spec.js]: Good catch. Would be great to report this as a bug 👍

**Review by talldan — APPROVED** [review]
Nice work, looks good to me!

File: test/e2e/playwright.config.ts (hunk 4 above)
Comment (kevin940726) [reply test/e2e/playwright.config.ts:63]: Yeah, it's the same as running in Chromium instead of Chrome or Edge. Another reason is that Safari is only available on macOS, but webkit is also available on Linux and Windows.

File: packages/block-editor/src/components/inserter-list-item/style.scss (hunk 1 above)
Comment (kevin940726) [reply packages/block-editor/src/components/inserter-list-item/style.scss:53]: I think we can keep this here since the added tests caught this bug. If we separate this into another PR then it wouldn't make the full picture (the corresponding test is in another PR).

File: .github/workflows/end2end-test-playwright.yml (hunk 5 above)
Comment (kevin940726) [reply .github/workflows/end2end-test-playwright.yml:42]: Here's the full output of the test results:

```
Running 8 tests using 1 worker

  ✓  1 [webkit] › editor/various/inserting-blocks.spec.js:21:2 › Inserting blocks (@firefox, @webkit) › inserts blocks by dragging and dropping from the global inserter (6s)
  ✓  2 [webkit] › editor/various/inserting-blocks.spec.js:100:2 › Inserting blocks (@firefox, @webkit) › cancels dragging blocks from the global inserter by pressing Escape (6s)
  ✓  3 [webkit] › editor/various/inserting-blocks.spec.js:157:2 › Inserting blocks (@firefox, @webkit) › inserts patterns by dragging and dropping from the global inserter (5s)
  ✓  4 [webkit] › editor/various/inserting-blocks.spec.js:227:2 › Inserting blocks (@firefox, @webkit) › cancels dragging patterns from the global inserter by pressing Escape (4s)
  -  5 [firefox] › editor/various/inserting-blocks.spec.js:21:2 › Inserting blocks (@firefox, @webkit) › inserts blocks by dragging and dropping from the global inserter
  ✓  6 [firefox] › editor/various/inserting-blocks.spec.js:100:2 › Inserting blocks (@firefox, @webkit) › cancels dragging blocks from the global inserter by pressing Escape (8s)
  -  7 [firefox] › editor/various/inserting-blocks.spec.js:157:2 › Inserting blocks (@firefox, @webkit) › inserts patterns by dragging and dropping from the global inserter
  ✓  8 [firefox] › editor/various/inserting-blocks.spec.js:227:2 › Inserting blocks (@firefox, @webkit) › cancels dragging patterns from the global inserter by pressing Escape (5s)
```

There is the browser/project name in front of each test case. Just that it won't show them if the terminal isn't wide enough. I think Playwright should at least show the project name and just truncate the middle part though, but that's up to them.

File: test/e2e/specs/editor/various/inserting-blocks.spec.js (hunk 6 above)
Comment (kevin940726) [reply test/e2e/specs/editor/various/inserting-blocks.spec.js]: Opened here: https://github.com/WordPress/gutenberg/issues/45183

File: .github/workflows/end2end-test-playwright.yml (hunk 2 above)
Comment (kevin940726) [reply .github/workflows/end2end-test-playwright.yml]: Opened https://github.com/WordPress/gutenberg/pull/45187.
